### PR TITLE
Clean up API requests

### PR DESCRIPTION
**Pull request by kntsoriano**
## Description
When we have API requests, we must do the ff:
1. Abort it when component unmounts.
2. Make sure that the component is still mounted when running side effects.

### CODE REVIEW

File: app/javascript/components/student/Editor.tsx [hunk 1]
```diff
@@ -233,25 +243,32 @@ export function Editor({
...
+        if (!isMountedRef.current) {
+          return
+        }
+
         const iteration = typecheck<Iteration>(json, 'iteration')
         location.assign(iteration.links.self)
       })
     },
-    [submission]
+    [controllerRef, isMountedRef, submission]
```
Comment (ErikSchierboom) [inline app/javascript/components/student/Editor.tsx:271]: Do we also need to add the following code here:

```
.finally(() => {
          controllerRef.current = undefined
        })
```

This is used in the above `fetchJSON` call.

File: app/javascript/components/student/Editor.tsx [hunk 2]
```diff
@@ -207,12 +209,20 @@ export function Editor({
         body: JSON.stringify({ files: files }),
       })
         .then((json: any) => {
+          if (!isMountedRef.current) {
```
Comment (ErikSchierboom) [inline app/javascript/components/student/Editor.tsx:212]: If we have this pattern at multiple places (where there is a `isMountedRef ` and `AbortController`), we could consider encoding this in some helper function at some point.

**Review by ErikSchierboom — APPROVED** [review]

File: app/javascript/components/student/Editor.tsx (hunk 1 above)
Comment (kntsoriano) [reply app/javascript/components/student/Editor.tsx:271]: Yes. We need to :) Thanks.

File: app/javascript/components/student/Editor.tsx (hunk 2 above)
Comment (kntsoriano) [reply app/javascript/components/student/Editor.tsx:212]: I agree! I created https://github.com/exercism/v3-website/issues/255 for this one.

Comment (kntsoriano) [reply app/javascript/components/student/Editor.tsx:212]: @SleeplessByte Have you got this as a custom hook?

**Pull request merged**